client/fingerprint/network_unix.go

```diff
@@ -152,99 +144,62 @@ func (f *NetworkFingerprint) linkSpeedEthtool(path, device string) int {
-					ip = v.IP.String()
-				}
-			}
-		}
+	if len(addrs) == 0 {
+		return "", errors.New(fmt.Sprintf("Interface %s has no IP address", intf.Name))
 	}
-
-	if net.ParseIP(ip) == nil {
-		return "", errors.New(fmt.Sprintf("could not parse IP address `%s`", ip))
+	if ip, _, err := net.ParseCIDR(addrs[0].String()); err == nil {
+		return ip.String(), nil
 	}
-
-	return ip, nil
+	return "", errors.New(fmt.Sprintf("Couldn't parse IP address for interface %s with addr %s", intf.Name, addrs[0].String()))
 }
 
-// ifConfig returns the IP Address for this node according to ifConfig, for the
-// specified device.
-func (f *NetworkFingerprint) ifConfig(device string) string {
-	ifConfigPath, _ := exec.LookPath("ifconfig")
-	if ifConfigPath == "" {
-		f.logger.Println("[WARN] fingerprint.network: ifconfig not found")
-		return ""
-	}
+func (f *NetworkFingerprint) isDeviceEnabled(intf *net.Interface) bool {
+	return intf.Flags&net.FlagUp != 0
+}
 
-	outBytes, err := exec.Command(ifConfigPath, device).Output()
-	if err != nil {
-		f.logger.Printf("[WARN] fingerprint.network: Error calling ifconfig (%s %s): %v", ifConfigPath, device, err)
-		return ""
+func (f *NetworkFingerprint) deviceHasIpAddress(intf *net.Interface) bool {
+	if addrs, err := intf.Addrs(); err == nil {
+		return len(addrs) > 0
 	}
+	return false
+}
 
-	// Parse out the IP address returned from ifconfig for this device
-	// Tested on Ubuntu, the matching part of ifconfig output for eth0 is like
-	// so:
-	//   inet addr:10.0.2.15  Bcast:10.0.2.255  Mask:255.255.255.0
-	// For OS X and en0, we have:
-	//  inet 192.168.0.7 netmask 0xffffff00 broadcast 192.168.0.255
-	output := strings.TrimSpace(string(outBytes))
-
-	// re is a regular expression, which can vary based on the OS
-	var re *regexp.Regexp
+func (n *NetworkFingerprint) isDeviceLoopBackOrPointToPoint(intf *net.Interface) bool {
+	return intf.Flags&(net.FlagLoopback|net.FlagPointToPoint) == 0
+}
 
-	if "darwin" == runtime.GOOS {
-		re = regexp.MustCompile("inet [0-9].+")
-	} else {
-		re = regexp.MustCompile("inet addr:[0-9].+")
+func (f *NetworkFingerprint) findInterfaces(deviceName string) ([]*net.Interface, error) {
+	var (
+		interfaces []*net.Interface
+		err        error
+	)
+	if deviceName != "" {
+		if intf, err := net.InterfaceByName(deviceName); err == nil {
+			interfaces = append(interfaces, intf)
+		}
+		return interfaces, err
 	}
-	args := strings.Split(re.FindString(output), " ")
 
-	var ip string
-	if len(args) > 1 {
-		ip = strings.TrimPrefix(args[1], "addr:")
+	var intfs []net.Interface
+
+	if intfs, err = net.Interfaces(); err != nil {
+		return nil, err
 	}
 
-	// validate what we've sliced out is a valid IP
-	if net.ParseIP(ip) == nil {
-		f.logger.Printf("[WARN] fingerprint.network: Unable to parse IP in output of '%s %s'", ifConfigPath, device)
-		return ""
+	for _, intf := range intfs {
+		if f.isDeviceEnabled(&intf) && f.isDeviceLoopBackOrPointToPoint(&intf) && f.deviceHasIpAddress(&intf) {
```

Comment:
Does this mean we _only_ find loopback and PPP interfaces? Is that what we want here?